content/into.md

```diff
@@ -0,0 +1,319 @@
+ - A type alias of a valid conversion target type.
+ - A match type that reduces to a valid conversion target type.
+ - An annotated type `T @ann` where `T` is a valid conversion target type.
+ - A refined type `T {...}` where `T` is a valid conversion target type.
+ - A union `T | U` of two valid conversion target types `T` and `U`.
+ - An intersection `T & U` of two valid conversion target types `T` and `U`.
+ - An instance of a type parameter that is explicitly instantiated to a valid conversion target type.
+
+
+Type parameters that are not fully instantiated do not count as valid conversion target types. For instance, consider:
+
+```scala
+  trait Token
+  class Keyword(str: String)
+  given Conversion[String, Keyword] = KeyWord(_)
+
+  List[into[Keyword]]("if", "then", "else")
+```
+This type-checks since the target type of the list elements is the type parameter of the `List.apply` method which is explicitly instantiated to `into[Keyword]`. On the other hand, if we continue the example as follows we get an error:
+```scala
+  val ifKW: into[Keyword] = "if"
+  val ys: List[into[Keyword]] = List(ifKW, "then", "else")
+```
+Here, the type variable of `List.apply` is not explicitly instantiated
+when we check the `List(...)` arguments (it is just upper-bounded by the target type `into[Keyword]`). This is not enough to allow
+implicit conversions on the second and third arguments.
+
+Subclasses of `into` classes or traits do not count as valid conversion target types. For instance, consider:
+
+```scala
+into trait T
+class C(x: Int) extends T
+given Conversion[Int, C] = C(_)
+
+def f(x: T) = ()
+def g(x: C) = ()
+f(1)      // ok
+g(1)      // error
+```
+The call `f("abc")` type-checks since `f`'s parameter type `T` is `into`.
+But the call `g("abc")` does not type-check since `g`'s parameter type `C` is not `into`. It does not matter that `C` extends a trait `T` that is `into`.
+
+
+## Why Two Different Schemes?
+
+Can we make do with just one scheme instead of two? In practice this would be difficult.
+
+Let's first take a look the `Expr` example, which uses into-as-a-constructor. Could it be rewritten to use into-as-a-modifier?
+This would mean we have to add `into` to the whole `Expr` enum. Adding it to just `Const` is not enough, since `Add` and `Neg` take `Expr` arguments, not `Const` arguments.
+
+But we might not always have permission to change the `Expr` enum. For instance, `Expr` could be defined in a lower level library without implicit conversions, but later we want to make `Expr` construction convenient by eliding `Const` wrappers in some higher-level library or application. With `into` constructors, this is easy: Define the implicit conversion and facade methods that construct `Expr` trees while taking `into[Expr]` parameters.
+With `into` modifiers there is no way to achieve the same.
```

Comment:
This could also be solved by allowing `into` as a modifier on type aliases. Then, the higher-level library is free to define:
```scala
into type intoExpr = Expr
```